<a href="https://colab.research.google.com/github/d4u0ul/Novo_projeto/blob/master/Spacy_com_NER2.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [53]:
!pip install -U spacy 
import spacy
!nvcc -V
spacy.require_gpu()
#spacy.required_gpu()

Requirement already up-to-date: spacy in /usr/local/lib/python3.6/dist-packages (2.3.2)
nvcc: NVIDIA (R) Cuda compiler driver
Copyright (c) 2005-2019 NVIDIA Corporation
Built on Sun_Jul_28_19:07:16_PDT_2019
Cuda compilation tools, release 10.1, V10.1.243


True

Numero dos arquivos escolhidos randomicamente para mostrar a acurácia do modelo no teste final


*import random*<br>
*for x in range (130):* <br>
  *print(random.randint(1,652))*<br>

198
295
382
608
187
80
127
62
506
529
59
315
628
35
274
222
452
58
613
629
324
73
345
42
113
616
180
61
446
416
651
165
172
614
82
70
95
348
288
347
455
605
137
208
114
431
39
91
534
28
120
43
122
332
108
538
405
309
305
521
450
286
87
357
518
605
3
325
364
250
113
639
429
70
488
514
456
209
38
595
49
157
126
335
522
324
102
53
97
333
534
589
219
65
475
173
179
264
625
378
494
127
629
645
345
341
567
147
338
317
637
310
157
396
565
400
246
407
577
378
528
640
325
217
581
538
159
550
191
14

4/5wFnEiljtx93u9C25s5VVRJhAFZ89xzhEBVeiIh8FIrDGGaYKUecSUk


In [54]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Recuperando os arquivos restantes na pasta

In [55]:
from os import chdir, getcwd, listdir
path= r'/content/drive/My Drive/ner_itens_medicamentos/'
listaNum=[]
listaDeArqEnt=[]
listaDeArqText=[]
chdir(path)
print(getcwd()) #print caminho 

for c in listdir(): #listdir são os arquivos dentro da pasta  
 if "text" in str(c) and "entities" not in str(c):
   listaNum.append(int(c[4:c.index(".")])) #add numero dos textos 



for n in listaNum: #listdir são os arquivos dentro 
  listaDeArqText.append("text"+str(n)+".txt") #add numero dos textos 
  listaDeArqEnt.append("text"+str(n)+"_entities.txt")



print(len(listaNum))


/content/drive/My Drive/ner_itens_medicamentos
534


In [56]:
for x in range(5):
#for x in range(len(listaNum)):
  print("o texto de numero %s é o %s e suas entidades estão em %s" %(listaNum[x],listaDeArqText[x],listaDeArqEnt[x]))

o texto de numero 207 é o text207.txt e suas entidades estão em text207_entities.txt
o texto de numero 390 é o text390.txt e suas entidades estão em text390_entities.txt
o texto de numero 588 é o text588.txt e suas entidades estão em text588_entities.txt
o texto de numero 45 é o text45.txt e suas entidades estão em text45_entities.txt
o texto de numero 139 é o text139.txt e suas entidades estão em text139_entities.txt



**Versão**
https://medium.com/@manivannan_data/how-to-train-ner-with-custom-training-data-using-spacy-188e0e508c6   

In [57]:
def impListaRange(lista):
  for x in range(len(lista)):
    print(str(x) +"-->> " + str(lista[x]))

def preparaTexto(texto):
  a=texto
  a=a.replace("art.","art#").replace("pag.","pag#").replace("arts.","arts#").replace("num.","num#").replace("fls.","fls#")
  a=a.replace(" . ", " ").replace("<permisc>"," ").replace("<idnum>","").replace("< < rmisc> rmisc> ","").replace("rmisc"," ").replace(">"," ").replace("<","").replace(".<", "<").replace("Código Texto","")
  a=a.split(". ")
  for i in range(len(a)):
    a[i]=a[i].replace("art#","art.").replace("pag#","pag.").replace("arts#","arts.").replace("num#","num.").replace("fls#","fls.")
    while "  " in a[i]:
      a[i]=a[i].replace("  "," ")
  return a
  
def impListaRange2(lista,ranges):
  for x in range(len(lista)):
    if x>=ranges:
      break
    print(lista[x],end="")
    if x!=ranges-1:
      print(",")

def impListaRange(lista):
  ranges=len(lista)
  for x in range(len(lista)):
    if x>=ranges:
      break
    print(lista[x],end="")
    if x!=ranges-1:
      print(",")


In [58]:
textoEntComp=[]
textoEntPars=[]
nEntPars=[]
ltrain_data=[]
train_data=[]
print(listaDeArqEnt[0])
print(listaDeArqEnt[1])
print(listaDeArqEnt[2])
sentencas=[]

for t in range(len(listaNum)):
  if t>300:
   break
  textoTxt=str((open(path+listaDeArqText[t],"r").read()))
 #textoTxt.append(()
  textoEntComp=open(path+listaDeArqEnt[t],"r")
  for x in textoEntComp:
    y=x[:x.index("\n")]
    z=y.split(",")
    nEntPars.append((int(z[0]),int(z[1]),"Item"))
  train_data.append((textoTxt,{'entities':list(nEntPars)}))
  print("appending text[%s]"%(t))  
  nEntPars.clear()
  
print('train_data carregado')

text207_entities.txt
text390_entities.txt
text588_entities.txt
appending text[0]
appending text[1]
appending text[2]
appending text[3]
appending text[4]
appending text[5]
appending text[6]
appending text[7]
appending text[8]
appending text[9]
appending text[10]
appending text[11]
appending text[12]
appending text[13]
appending text[14]
appending text[15]
appending text[16]
appending text[17]
appending text[18]
appending text[19]
appending text[20]
appending text[21]
appending text[22]
appending text[23]
appending text[24]
appending text[25]
appending text[26]
appending text[27]
appending text[28]
appending text[29]
appending text[30]
appending text[31]
appending text[32]
appending text[33]
appending text[34]
appending text[35]
appending text[36]
appending text[37]
appending text[38]
appending text[39]
appending text[40]
appending text[41]
appending text[42]
appending text[43]
appending text[44]
appending text[45]
appending text[46]
appending text[47]
appending text[48]
appending text[4

In [59]:
impListaRange2(train_data,10) #impressão padrão para entrada no spacy

('processo <permisc> 1o grau: https://pje.tjpe.jus.br/1g/painel/painel_usuario/documentohtml.seam... intimacao do <permisc> para cumprir a decisao judicial, visto ser ele o responsavel na regiao pelo fornecimento de medicamentos a populacao. ademais, restou demonstrado nos autos que o ilustre magistrado, ato continuo, determinou a citacao do estado de sao paulo para os termos da acao, bem como do teor do despacho que deferiu aliminar, conforme se depreende da carta precatoria juntada as fls.52. nesse contexto, evidente que nao houve violacao ao principio do contraditorio e, por conseguinte, inexiste a nulidade aventada." (tjsp - <idnum> <permisc> / prestacao <permisc> publico; relator(a): <permisc> ; comarca: ilha solteira; orgao julgador: 12a camara de direito publico; data do julgamento: 24/06/ <idnum> ; data de registro: 06/08/ <idnum> ; outros numeros: <idnum> [1]) assinado eletronicamente por: <permisc> dias ribeiro de lavigne https://pje.tjpe.jus.br:443/1g/processo/consultadocume

In [60]:
#Descobrindo o nome dos Itens
nomeDosMedicamentos=[]
cont=0
for i in range(len(train_data)):
  print("###################################################################################################################################################################################################################################")
  print("Analisando o texto de numero %s que é o %s e que suas entidades estão em %s, temos" %(str(cont+1),listaDeArqText[i],listaDeArqEnt[i]))
  print("%s: %s"%(listaDeArqText[i],str(train_data[i][0])))
  listaEntidades=list((train_data[i][1].get("entities")))
  print("Lista de entidades do texto %s de %i caracteres são %s"%(listaDeArqText[i],len(str(train_data[i][0])),listaEntidades))
  listaMed=list([])
  for j in range(len(listaEntidades)):
    print()
    print("A palavra procurada em %s que "%(str(listaEntidades[j])),end=" ")
    print("deve estar entre %i" %(listaEntidades[j][0]),end=" ")
    print(listaEntidades[j][1],end=" é a palavra ")
    print(train_data[i][0][listaEntidades[j][0]:listaEntidades[j][1]])
    listaMed.append(train_data[i][0][listaEntidades[j][0]:listaEntidades[j][1]])
            #train_data.append((textoTxt,{'entities':list(nEntPars)}))
    print()
    for k in range(len(listaMed)):
      if k==0:continue
      if listaMed[k]==listaMed[k-1]:
        del listaMed[k-1]
  nomeDosMedicamentos.append((listaDeArqText[i],{'Medicamentos':list(listaMed)}))
   # for k in 
    
   # print(j,end=" -> ")
    #print(train_data[i][0][j[0][0listaEntidades[j][1]]])
  #   sentecas.append([list(preparaTexto(i[0][0]))][j])
  cont+=1


print("")
a=train_data[0][1]
b=train_data[1][1]
c=train_data[2][1]
d=train_data[3][1]
e=train_data[4][1]
f=train_data[5][1]

#print(a[8071:8082])
#print(len(a))
print(a)
print(train_data[1][1].get("entities"))
print(c['entities'])
print(d['entities'][0][0])
print(e['entities'])
print(f['entities'])

###################################################################################################################################################################################################################################
Analisando o texto de numero 1 que é o text207.txt e que suas entidades estão em text207_entities.txt, temos
text207.txt: processo <permisc> 1o grau: https://pje.tjpe.jus.br/1g/painel/painel_usuario/documentohtml.seam... intimacao do <permisc> para cumprir a decisao judicial, visto ser ele o responsavel na regiao pelo fornecimento de medicamentos a populacao. ademais, restou demonstrado nos autos que o ilustre magistrado, ato continuo, determinou a citacao do estado de sao paulo para os termos da acao, bem como do teor do despacho que deferiu aliminar, conforme se depreende da carta precatoria juntada as fls.52. nesse contexto, evidente que nao houve violacao ao principio do contraditorio e, por conseguinte, inexiste a nulidade aventada." (tjsp - <idnum> <permis

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



In [61]:
#teste lista de nomes de medicamentos
impListaRange2(nomeDosMedicamentos,len(nomeDosMedicamentos))

('text207.txt', {'Medicamentos': ['beclometasona']}),
('text390.txt', {'Medicamentos': ['ofev 150mg ']}),
('text588.txt', {'Medicamentos': ['ressonancia magnetica de cranio com contraste e sedacao']}),
('text45.txt', {'Medicamentos': ['somatropina 12 ui']}),
('text139.txt', {'Medicamentos': ['pertuzumabe']}),
('text287.txt', {'Medicamentos': ['temozolomida']}),
('text413.txt', {'Medicamentos': ['cadeira de rodas']}),
('text425.txt', {'Medicamentos': ['ritalina']}),
('text548.txt', {'Medicamentos': ['ressonancia magnetica de cranio com contraste e sedacao']}),
('text298.txt', {'Medicamentos': ['xeralto 20 mg']}),
('text75.txt', {'Medicamentos': ['fabrazyme']}),
('text89.txt', {'Medicamentos': ['lucentis']}),
('text131.txt', {'Medicamentos': ['cardiaca']}),
('text245.txt', {'Medicamentos': ['infliximabe', 'infliximab)']}),
('text266.txt', {'Medicamentos': ['oxibutinina']}),
('text372.txt', {'Medicamentos': ['rivaroxabana']}),
('text118.txt', {'Medicamentos': ['nivolumabe']}),
('text427.t

In [62]:
#dividir os textos
train_data2=[]
for i in range(len(train_data)):
  sentencas=list(preparaTexto(train_data[i][0]))
  
  for j in sentencas:
    
    for k in (nomeDosMedicamentos[i][1].get("Medicamentos")):
      indices=list([])
      if k in j:
        print(k)
        # nEntPars.append((int(z[0]),int(z[1]),"Item"))
        #train_data.append((textoTxt,{'entities':list(nEntPars)}))
        indices.append((int(j.index(k)),int(j.index(k))+len(k),"Item"))
        print(indices,end=" ->")
        print(j[(int(j.index(k))):(int(j.index(k)))+len(k)])
        train_data2.append((j,{'entities':list(indices)}))   
        
      else:
        #print("nome não encontrado na sentenca")
        continue
       
print(len(train_data2))  
      

beclometasona
[(174, 187, 'Item')] ->beclometasona
ofev 150mg 
[(502, 513, 'Item')] ->ofev 150mg 
ressonancia magnetica de cranio com contraste e sedacao
[(188, 243, 'Item')] ->ressonancia magnetica de cranio com contraste e sedacao
ressonancia magnetica de cranio com contraste e sedacao
[(109, 164, 'Item')] ->ressonancia magnetica de cranio com contraste e sedacao
ressonancia magnetica de cranio com contraste e sedacao
[(80, 135, 'Item')] ->ressonancia magnetica de cranio com contraste e sedacao
somatropina 12 ui
[(522, 539, 'Item')] ->somatropina 12 ui
somatropina 12 ui
[(376, 393, 'Item')] ->somatropina 12 ui
somatropina 12 ui
[(389, 406, 'Item')] ->somatropina 12 ui
pertuzumabe
[(662, 673, 'Item')] ->pertuzumabe
pertuzumabe
[(177, 188, 'Item')] ->pertuzumabe
pertuzumabe
[(183, 194, 'Item')] ->pertuzumabe
temozolomida
[(883, 895, 'Item')] ->temozolomida
temozolomida
[(355, 367, 'Item')] ->temozolomida
temozolomida
[(105, 117, 'Item')] ->temozolomida
cadeira de rodas
[(310, 326, 'Ite

In [63]:
print(train_data2[0][1])
print(train_data2[1][1])
print(train_data2[2][1])
impListaRange2(train_data2,5)
#{'entities': [(2794, 2802, 'Item'), (22366, 22374, 'Item'), (29318, 29326, 'Item')]}

print(train_data2)

{'entities': [(174, 187, 'Item')]}
{'entities': [(502, 513, 'Item')]}
{'entities': [(188, 243, 'Item')]}
('dos fatos e fundamentos juridicos o autor em virtude de estar acometido por doenca pulmonar obstrutiva cronica ( ), necessita utilizar a combinacao dos farmacos formoterol + beclometasona 12/400 (02 vezes por dia) e tiotropico 2,5mg (02 jatos por dia),conforme relatorio medico, a fim de evitar insuficiencia respiratoria e, consequentemente, o risco de obito', {'entities': [(174, 187, 'Item')]}),
(' /50 e /83,uma vez nao dispondo de meios para prover as despesas do processo sem prejuizo do sustento proprio ouda familia; c) seja o estado de pernambuco citado, por meio de perm sc (art. 247, perm sc ), na pessoade seu representante legal, para contestar o feito, sob pena de aplicacao dos efeitos da revelia; d) perm sc o deferimento da perm sc , conforme faculta o inciso e perm sc do artigo294 do , condenando a empresa re, a cobrir o ; , para assegurar um tratamento de forma adequada, 

In [64]:
#import pickle
#pickleIA = pickle.dumps(train_data2)
#
## salvar o pickleIA no arquivo pickleSpacy.pkl
#with open('pickleSpacy.pkl', 'wb') as file:
#    pickle.dump(train_data2, file)

In [65]:
##test=(open(path+'pickleSpacy.pkl',"r").read())
##test2= pickle.loads(test)
##print(test2)
#
## Carregar modelo
#with open(path+'pickleSpacy.pkl', 'rb') as f:
#    test = pickle.load(f)
##print(pickle.loads(f))

*Training*

You can start the training once you completed the second step. Just copy the text and paste into TRAIN_DATA variable in train.py. Run the training…

In [ ]:
import spacy
import random


TRAIN_DATA = train_data2


def train_spacy(data,iterations):
    TRAIN_DATA = data
    nlp = spacy.blank('en')  # create blank Language class
    # create the built-in pipeline components and add them to the pipeline
    # nlp.create_pipe works for built-ins that are registered with spaCy
    if 'ner' not in nlp.pipe_names:
        ner = nlp.create_pipe('ner')
        nlp.add_pipe(ner, last=True)
       

    # add labels
    for _, annotations in TRAIN_DATA:
         for ent in annotations.get('entities'):
            ner.add_label(ent[2])

    # get names of other pipes to disable them during training
    other_pipes = [pipe for pipe in nlp.pipe_names if pipe != 'ner']
    with nlp.disable_pipes(*other_pipes):  # only train NER
        optimizer = nlp.begin_training()
        for itn in range(iterations):
            print("Statring iteration " + str(itn))
            random.shuffle(TRAIN_DATA)
            losses = {}
            for text, annotations in TRAIN_DATA:
                nlp.update(
                    [text],  # batch of texts
                    [annotations],  # batch of annotations
                    drop=0.2,  # dropout - make it harder to memorise data
                    sgd=optimizer,  # callable to update weights
                    losses=losses)
            print(losses)
    return nlp


prdnlp = train_spacy(TRAIN_DATA, 100)

# Save our trained Model
#modelfile = input("Enter your Model Name: ")
prdnlp.to_disk("Poli_IA_SES")



/usr/local/lib/python3.6/dist-packages/spacy/language.py:639: UserWarning: [W033] Training a new parser or NER using a model with no lexeme normalization table. This may degrade the performance of the model to some degree. If this is intentional or the language you're using doesn't have a normalization table, please ignore this warning. If this is surprising, make sure you have the spacy-lookups-data package installed. The languages with lexeme normalization tables are currently: da, de, el, en, id, lb, pt, ru, sr, ta, th.
  **kwargs
/usr/local/lib/python3.6/dist-packages/spacy/language.py:482: UserWarning: [W030] Some entities could not be aligned in the text "sendo assim, o autor necessita mudar o seu tratame..." with entities "[(90, 106, 'Item')]". Use `spacy.gold.biluo_tags_from_offsets(nlp.make_doc(text), entities)` to check the alignment. Misaligned entities ('-') will be ignored during training.
  gold = GoldParse(doc, **gold)


Statring iteration 0


/usr/local/lib/python3.6/dist-packages/spacy/language.py:482: UserWarning: [W030] Some entities could not be aligned in the text " 16/17 e) a citacao dos reus para, querendo, respo..." with entities "[(323, 334, 'Item')]". Use `spacy.gold.biluo_tags_from_offsets(nlp.make_doc(text), entities)` to check the alignment. Misaligned entities ('-') will be ignored during training.
  gold = GoldParse(doc, **gold)
/usr/local/lib/python3.6/dist-packages/spacy/language.py:482: UserWarning: [W030] Some entities could not be aligned in the text " assinado eletronicamente por: elo sa perm sc rodr..." with entities "[(354, 379, 'Item')]". Use `spacy.gold.biluo_tags_from_offsets(nlp.make_doc(text), entities)` to check the alignment. Misaligned entities ('-') will be ignored during training.
  gold = GoldParse(doc, **gold)
/usr/local/lib/python3.6/dist-packages/spacy/language.py:482: UserWarning: [W030] Some entities could not be aligned in the text "entrementes, a solicitacao fora negada sob o argum..

{'ner': 1609.2805664219904}
Statring iteration 1


/usr/local/lib/python3.6/dist-packages/spacy/language.py:482: UserWarning: [W030] Some entities could not be aligned in the text " fatos o autor, atualmente com 81 anos de idade, e..." with entities "[(293, 311, 'Item')]". Use `spacy.gold.biluo_tags_from_offsets(nlp.make_doc(text), entities)` to check the alignment. Misaligned entities ('-') will be ignored during training.
  gold = GoldParse(doc, **gold)
  

In [ ]:
import spacy
nlp = spacy.load('Poli_IA_SES')

In [ ]:
from os import chdir, getcwd, listdir
path= r'/content/drive/My Drive/ner_itens_medicamentos/TesteFinal/'
listaNumTF=[]
listaDeArqEntTF=[]
listaDeArqTextTF=[]
chdir(path)
print(getcwd()) #print caminho 

for c in listdir(): #listdir são os arquivos dentro da pasta  
 if "text" in str(c) and "entities" not in str(c):
   listaNumTF.append(int(c[4:c.index(".")])) #add numero dos textos 



for n in listaNumTF: #listdir são os arquivos dentro 
  listaDeArqTextTF.append("text"+str(n)+".txt") #add numero dos textos 
  listaDeArqEntTF.append("text"+str(n)+"_entities.txt")

#listadetestes="198 295 382 608 187 80 127 62 506 529 59 315 628 35 274 222 452 58 613 629 324 73 345 42 113 616 180 61 446 416 651 165 172 614 82 70 95 348 288 347 455 605 137 208 114 431 39 91 534 28 120 43 122 332 108 538 405 309 305 521 450 286 87 357 518 605 3 325 364 250 113 639 429 70 488 514 456 209 38 595 49 157 126 335 522 324 102 53 97 333 534 589 219 65 475 173 179 264 625 378 494 127 629 645 345 341 567 147 338 317 637 310 157 396 565 400 246 407 577 378 528 640 325 217 581 538 159 550 191 14"
#listadetestes=listadeteste.replace(" ",",")


print(len(listaNumTF))

In [ ]:
for x in range(5):
#for x in range(len(listaNum)):
  print("o texto de numero %s é o %s e suas entidades estão em %s" %(listaNumTF[x],listaDeArqTextTF[x],listaDeArqEntTF[x]))

In [ ]:
data=[]
for t in range(len(listaNumTF)):
  #if t>650:
   # break
  textoTxtTF=(open(path+listaDeArqTextTF[t],"r").read())
 #textoTxt.append(()
  textoEntCompTF=open(path+listaDeArqEntTF[t],"r")
  for x in textoEntCompTF:
    y=x[:x.index("\n")]
    z=y.split(",")
    nEntPars.append((int(z[0]),int(z[1]),"texto"))
  data.append((textoTxtTF,{'entities':list(nEntPars)}))
  nEntPars.clear()
  
#print(data)
print(data[0][0])


In [ ]:
#Test your text
#test_text = input("Enter your testing text: ")
peticao = prdnlp(data[0][0])
for ent in peticao.ents:
    print(ent.text, ent.start_char, ent.end_char, ent.label_)

In [ ]:
#peticao = nlp()

In [ ]:
from spacy import displacy
displacy.render(peticao, style = 'ent', jupyter = True) #imprime o texto com um render nas palavras rotuladas

In [ ]:
peticao.ents #verificando as entidades (palavras)

In [ ]:
for ent in peticao.ents: #imprime as palavras junto com seus rotulos
  print(ent.text, ent.label_)

In [ ]:
[token for token in peticao] #imprime token por token (item do texto)

In [ ]:
[(token.orth_, token.pos_) for token in peticao] #Analise de classes gramaticais

Agora
Já está no github

Próximos passos
Refinar o código
Estudar formas mais eficientes de executar o código

Pensando alto
Implementar um sistema web front-end básico 
